File: go/test/book_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/fake-json-sdk/go"
	"github.com/voxgig-sdk/fake-json-sdk/go/core"

	vs "github.com/voxgig-sdk/fake-json-sdk/go/utility/struct"
)

func TestBookEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Book(nil)
		if ent == nil {
			t.Fatal("expected non-nil BookEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"book": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Book(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Book(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := bookBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "book." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set FAKE_JSON_TEST_BOOK_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		bookRef01Ent := client.Book(nil)
		bookRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "book"}, setup.data), "book_ref01"))

		bookRef01DataResult, err := bookRef01Ent.Create(bookRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		bookRef01Data = core.ToMapAny(entityData(bookRef01DataResult))
		if bookRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if bookRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		bookRef01Match := map[string]any{}

		bookRef01ListResult, err := bookRef01Ent.List(bookRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		bookRef01List, bookRef01ListOk := bookRef01ListResult.([]any)
		if !bookRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", bookRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(bookRef01List), map[string]any{"id": bookRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		bookRef01DataUp0Up := map[string]any{
			"id": bookRef01Data["id"],
		}

		bookRef01MarkdefUp0Name := "author"
		bookRef01MarkdefUp0Value := fmt.Sprintf("Mark01-book_ref01_%d", setup.now)
		bookRef01DataUp0Up[bookRef01MarkdefUp0Name] = bookRef01MarkdefUp0Value

		bookRef01ResdataUp0Result, err := bookRef01Ent.Update(bookRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		bookRef01ResdataUp0 := core.ToMapAny(entityData(bookRef01ResdataUp0Result))
		if bookRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if bookRef01ResdataUp0["id"] != bookRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if bookRef01ResdataUp0[bookRef01MarkdefUp0Name] != bookRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", bookRef01MarkdefUp0Name, bookRef01ResdataUp0[bookRef01MarkdefUp0Name])
		}

		// LOAD
		bookRef01MatchDt0 := map[string]any{
			"id": bookRef01Data["id"],
		}
		bookRef01DataDt0Loaded, err := bookRef01Ent.Load(bookRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		bookRef01DataDt0LoadResult := core.ToMapAny(entityData(bookRef01DataDt0Loaded))
		if bookRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if bookRef01DataDt0LoadResult["id"] != bookRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		bookRef01MatchRm0 := map[string]any{
			"id": bookRef01Data["id"],
		}
		_, err = bookRef01Ent.Remove(bookRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		bookRef01MatchRt0 := map[string]any{}

		bookRef01ListRt0Result, err := bookRef01Ent.List(bookRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		bookRef01ListRt0, bookRef01ListRt0Ok := bookRef01ListRt0Result.([]any)
		if !bookRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", bookRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(bookRef01ListRt0), map[string]any{"id": bookRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func bookBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "book", "BookTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read book test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse book test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"book01", "book02", "book03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("FAKE_JSON_TEST_BOOK_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"FAKE_JSON_TEST_BOOK_ENTID": idmap,
		"FAKE_JSON_TEST_LIVE":      "FALSE",
		"FAKE_JSON_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["FAKE_JSON_TEST_BOOK_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["FAKE_JSON_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewFakeJsonSDK(core.ToMapAny(mergedOpts))
	}

	live := env["FAKE_JSON_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["FAKE_JSON_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/BookEntityTest.php
<?php
declare(strict_types=1);

// Book entity test

require_once __DIR__ . '/../fakejson_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class BookEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = FakeJsonSDK::test(null, null);
        $ent = $testsdk->Book(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "book" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = FakeJsonSDK::test($seed, null);
        $seen = iterator_to_array($base->Book(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = FakeJsonConfig::shared_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = FakeJsonSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->Book(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = book_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list", "update", "load", "remove"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "book." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set FAKE_JSON_TEST_BOOK_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $book_ref01_ent = $client->Book(null);
        $book_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.book"), "book_ref01"));

        $book_ref01_data_result = $book_ref01_ent->create($book_ref01_data, null);
        $book_ref01_data = Helpers::to_map(is_object($book_ref01_data_result) && method_exists($book_ref01_data_result, 'data_get') ? $book_ref01_data_result->data_get() : $book_ref01_data_result);
        $this->assertNotNull($book_ref01_data);
        $this->assertNotNull($book_ref01_data["id"]);

        // LIST
        $book_ref01_match = [];

        $book_ref01_list_result = $book_ref01_ent->list($book_ref01_match, null);
        $this->assertIsArray($book_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($book_ref01_list_result),
            ["id" => $book_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

        // UPDATE
        $book_ref01_data_up0_up = [
            "id" => $book_ref01_data["id"],
        ];

        $book_ref01_markdef_up0_name = "author";
        $book_ref01_markdef_up0_value = "Mark01-book_ref01_" . $setup["now"];
        $book_ref01_data_up0_up[$book_ref01_markdef_up0_name] = $book_ref01_markdef_up0_value;

        $book_ref01_resdata_up0_result = $book_ref01_ent->update($book_ref01_data_up0_up, null);
        $book_ref01_resdata_up0 = Helpers::to_map(is_object($book_ref01_resdata_up0_result) && method_exists($book_ref01_resdata_up0_result, 'data_get') ? $book_ref01_resdata_up0_result->data_get() : $book_ref01_resdata_up0_result);
        $this->assertNotNull($book_ref01_resdata_up0);
        $this->assertEquals($book_ref01_resdata_up0["id"], $book_ref01_data_up0_up["id"]);
        $this->assertEquals($book_ref01_resdata_up0[$book_ref01_markdef_up0_name], $book_ref01_markdef_up0_value);

        // LOAD
        $book_ref01_match_dt0 = [
            "id" => $book_ref01_data["id"],
        ];
        $book_ref01_data_dt0_loaded = $book_ref01_ent->load($book_ref01_match_dt0, null);
        $book_ref01_data_dt0_load_result = Helpers::to_map(is_object($book_ref01_data_dt0_loaded) && method_exists($book_ref01_data_dt0_loaded, 'data_get') ? $book_ref01_data_dt0_loaded->data_get() : $book_ref01_data_dt0_loaded);
        $this->assertNotNull($book_ref01_data_dt0_load_result);
        $this->assertEquals($book_ref01_data_dt0_load_result["id"], $book_ref01_data["id"]);

        // REMOVE
        $book_ref01_match_rm0 = [
            "id" => $book_ref01_data["id"],
        ];
        $book_ref01_ent->remove($book_ref01_match_rm0, null);

        // LIST
        $book_ref01_match_rt0 = [];

        $book_ref01_list_rt0_result = $book_ref01_ent->list($book_ref01_match_rt0, null);
        $this->assertIsArray($book_ref01_list_rt0_result);

        $not_found_item = sdk_select(
            Runner::entity_list_to_data($book_ref01_list_rt0_result),
            ["id" => $book_ref01_data["id"]]);
        $this->assertEmpty($not_found_item);

    }
}

function book_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/book/BookTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = FakeJsonSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["book01", "book02", "book03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("FAKE_JSON_TEST_BOOK_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "FAKE_JSON_TEST_BOOK_ENTID" => $idmap,
        "FAKE_JSON_TEST_LIVE" => "FALSE",
        "FAKE_JSON_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["FAKE_JSON_TEST_BOOK_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["FAKE_JSON_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new FakeJsonSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["FAKE_JSON_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["FAKE_JSON_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
